http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/AwsCrtResponseBodyPublisher.java

```diff
@@ -0,0 +1,298 @@
+    private final int windowSize;
+    private final AtomicBoolean isCancelled = new AtomicBoolean(false);
+    private final AtomicBoolean areNativeResourcesReleased = new AtomicBoolean(false);
+    private final AtomicBoolean isSubscriptionComplete = new AtomicBoolean(false);
+    private final AtomicBoolean queueComplete = new AtomicBoolean(false);
+    private final AtomicInteger mutualRecursionDepth = new AtomicInteger(0);
+    private final AtomicInteger queuedBytes = new AtomicInteger(0);
+    private final AtomicReference<Subscriber<? super ByteBuffer>> subscriberRef = new AtomicReference<>(null);
+    private final Queue<byte[]> queuedBuffers = new ConcurrentLinkedQueue<>();
+    private final AtomicReference<Throwable> error = new AtomicReference<>(null);
+
+    /**
+     * Adapts a streaming AWS CRT Http Response Body to a Publisher<ByteBuffer>
+     * @param stream The AWS CRT Http Stream for this Response
+     * @param windowSize The max allowed bytes to be queued. The sum of the sizes of all queued ByteBuffers should
+     *                   never exceed this value.
+     */
+    public AwsCrtResponseBodyPublisher(HttpClientConnection connection, HttpStream stream,
+                                       CompletableFuture<Void> responseComplete, int windowSize) {
+        Validate.notNull(connection, "HttpConnection must not be null");
+        Validate.notNull(stream, "Stream must not be null");
+        Validate.notNull(responseComplete, "Stream must not be null");
+        Validate.isPositive(windowSize, "windowSize must be > 0");
+        this.connection = connection;
+        this.stream = stream;
+        this.responseComplete = responseComplete;
+        this.windowSize = windowSize;
+    }
+
+    /**
+     * Method for the users consuming the Http Response Body to register a subscriber.
+     * @param subscriber The Subscriber to register.
+     */
+    @Override
+    public void subscribe(Subscriber<? super ByteBuffer> subscriber) {
+        Validate.notNull(subscriber, "Subscriber must not be null");
+
+        boolean wasFirstSubscriber = subscriberRef.compareAndSet(null, subscriber);
+
+        if (!wasFirstSubscriber) {
+            log.error(() -> "Only one subscriber allowed");
+
+            // onSubscribe must be called first before onError gets called, so give it a do-nothing Subscription
+            subscriber.onSubscribe(new Subscription() {
+                @Override
+                public void request(long n) {
+                    // This is a dummy implementation to allow the onError call
+                }
+
+                @Override
+                public void cancel() {
+                    // This is a dummy implementation to allow the onError call
+                }
+            });
+            subscriber.onError(new IllegalStateException("Only one subscriber allowed"));
+        } else {
+            subscriber.onSubscribe(new AwsCrtResponseBodySubscription(this));
```

Comment:
I don't quite follow the change here.  moving request() and publishToSubscribers() and their downstreams would be a really big refactor.  What about folding the logic of Subscription's request() into a single method on the publisher (rather than call 4 methods on the publisher)?

Reply:
Yeah, ideally I'd expect most of the logic to be in `Subscription` class rather than `Publisher` class. How about making `AwsCrtResponseBodySubscription` an internal class of `AwsCrtResponseBodyPublisher`? Seems cleaner that way to me.